### PR TITLE
Add processing of underlying HTTP level errors.

### PR DESCRIPTION
**Pull request by ekoval**
##### What is done:
- Add processing of underlying HTTP level errors
- Corrected processing of Authorization Failed case.
- Added test for fixed case.
- Added ChangeLog
- Bumped version.

https://github.com/yola/production/issues/2180

### CODE REVIEW

File: ultradns/client.py [hunk 1]
```diff
@@ -38,9 +41,12 @@ def _handle_error(self, response):
         if isinstance(json_body, list):
             error_code = json_body[0]['errorCode']
             error_msg = json_body[0]['errorMessage']
-        else:
+        elif isinstance(json_body, dict):
             error_code = json_body['errorCode']
             error_msg = json_body['errorMessage']
+        else:
+            error_code = ERR_HTTP
+            error_msg = 'HTTP-level error. Status code: %s; response body: %s' % (response.status_code, response.content)
```
Comment (snitch) [inline ultradns/client.py]: E501 line too long (121 > 79 characters)

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: ultradns/client.py [hunk 2]
```diff
@@ -16,13 +17,15 @@
 ERR_RECORDS_NOT_FOUND = 70002
 ERR_RECORD_ALREADY_EXISTS = 2111
 ERR_RECORD_NOT_FOUND = 56001
+ERR_HTTP = 111222333
```
Comment (stefanor) [inline ultradns/client.py]: Did we just make that up?

File: ultradns/client.py [hunk 3]
```diff
@@ -38,9 +41,14 @@ def _handle_error(self, response):
...
             error_code = json_body[0]['errorCode']
             error_msg = json_body[0]['errorMessage']
-        else:
+        elif isinstance(json_body, dict):
             error_code = json_body['errorCode']
             error_msg = json_body['errorMessage']
+        else:
+            error_code = ERR_HTTP
+            error_msg = """
+HTTP-level error. HTTP code: %s; response body: %s""" % (
```
Comment (sgalaburda) [inline ultradns/client.py]: The way of string wrapping here looks a bit strange.  Do we really need EOL at the beginning?

File: ultradns/client.py [hunk 4]
```diff
@@ -75,10 +83,12 @@ def is_auth_token_expired(self, response):
         """Return True if response says that the authentication token
         is expired.
         """
-        json_body = response.json()
-        return (response.status_code == requests.codes.UNAUTHORIZED and
-                isinstance(json_body, dict) and
-                json_body.get('errorCode') == ERR_CODE_AUTH_TOKEN_EXPIRED)
+        if response.status_code == requests.codes.UNAUTHORIZED:
+            json_body = response.json()
```
Comment (sgalaburda) [inline ultradns/client.py]: In multiple places in the code you pass response as an argument and then call `response.json()`.  I think we can get response json once and pass response_data dict to such methods.

File: ultradns/exceptions.py [hunk 5]
```diff
@@ -34,6 +34,10 @@ class RecordsNotFoundError(DNSError):
     pass
 
 
+class HTTPLevelError(DNSError):
```
Comment (sgalaburda) [inline ultradns/exceptions.py]: HTTPError?

File: ultradns/client.py (hunk 3 above)
Comment (deonbredenhann) [inline ultradns/client.py]: @sgalaburda  I usually do the following:
``` python
            error_msg = (
                'HTTP-level error. HTTP code: %s; response body: %s' % (
                    response.status_code, response.content))
```
or
``` python
            error_msg = ('HTTP-level error. HTTP code: %s; response '
                         'body: %s' % (response.status_code, response.content))
```

File: ultradns/client.py (hunk 2 above)
Comment (ekoval) [inline ultradns/client.py]: We need some error code for such kind of errors. DNS-level errors have their codes assigned by UltraDNS. This one is new.

File: ultradns/client.py (hunk 4 above)
Comment (ekoval) [inline ultradns/client.py]: Right. I just wanted to have as much context in it is possible, so I pass whole response object.

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: tests/client_tests.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        requests_mock.return_value = response_mock
+
+
+class TestCreatePrimaryZone(ClientTestCase):
+
+    @patch.object(ultradns.client.UltraDNSAuthentication, 'is_authenticated')
+    @patch.object(ultradns.client.requests, 'request')
+    def test_http_error_is_raised(self, post_mock, auth_mock):
+        self._mock_response(post_mock, '', 504, content='')
+        auth_mock.return_value=True
```
Comment (snitch) [inline tests/client_tests.py]: E225 missing whitespace around operator

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# [Changelog](https://github.com/yola/ultradns/releases)
+
+## 0.1.2
+* Add processing of HTTP-level errors.
```
Comment (RayeN) [inline CHANGELOG.md]: `Added` ?

File: ultradns/client.py [hunk 8]
```diff
@@ -38,9 +41,14 @@ def _handle_error(self, response):
...
             error_msg = json_body[0]['errorMessage']
-        else:
+        elif isinstance(json_body, dict):
             error_code = json_body['errorCode']
             error_msg = json_body['errorMessage']
+        else:
+            error_code = ERR_HTTP
+            error_msg = (
+                'HTTP-level error. HTTP code: %s; response body: %s' % (
+                    response.status_code, response.content))
```
Comment (RayeN) [inline ultradns/client.py]: This is already a separate case - so, you can just raise `HTTPError` directly here

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by RayeN** [comment]
:+1: Looks good to me.

**Pull request merged**